CP-20432: MxGPU: Load kernel module; add VF pcis to DB

thomassa opened the pull request:
In a new function Xapi_pgpu.mxgpu_vf_setup (to be called on starting a
VM that uses a vGPU backed by an MxGPU physical device) we:
* ensure the kernel module (driver) is loaded,
* add database entries for the new PCI devices that appear for the
  Virtual Functions, and
* link each of them to its Physical Function.

Incidentally add some utility functions:
* Picops.is_bdf_format
* Xapi_pci.get_local_pcis_and_records
* Xapi_pci.get_local_pci_refs

thomassa commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -307,3 +354,25 @@ let disable_dom0_access ~__context ~self =
+  ignore (Forkhelpers.execute_command_get_output !Xapi_globs.modprobe_path ["gim"]);
+  (* Update the gpus even if the module was present already, in case it was
+   * already loaded before xapi was (re)started. *)
+  Xapi_pci.update_pcis ~__context ~host:localhost;
```
Should this call be wrapped in `switched_sync`, the way we do it in `dbsync_slave.ml`?

thomassa replied:
@jonludlam says no.

robhoes replied:
Me too.

robhoes commented:
@thomassa Looks like you need a rebase.

robhoes commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -68,15 +68,62 @@ let sync_pci_hidden ~__context ~pgpu ~pci =
+    let linktarget = Unix.readlink path in
+    (* Expect linktarget to look like "../0000:8c:01.0" *)
+    (* If it does then cut off "../" and look up Db ref with that pci_id (i.e. address), and create a link in DB *)
+    if (try Scanf.sscanf linktarget "../%_4x:%_2x:%_2x:%_1x%!" true
```
I think it would be safer to not make any assumptions about the path other than the basename. The kernel or driver, or even `readlink`, might change such that an absolute path is returned, or one that looks slightly different in some other way. We could instead find the PCI address by `Filename.basename`.

robhoes replied:
We already have similar scanf function in pciops.ml. Can we reuse and/or extend them?

robhoes commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -68,15 +68,62 @@ let sync_pci_hidden ~__context ~pgpu ~pci =
+    if (try Scanf.sscanf linktarget "../%_4x:%_2x:%_2x:%_1x%!" true
+        with Scanf.Scan_failure _ | Failure _ | End_of_file -> false
+    ) then (
+      let physfn_addr = String.sub linktarget 3 12 in
```
As above. I'd change this to

    let physfn_addr = Filename.basename linktarget in

and move it above the scanf, so it can be used there.

robhoes commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -68,15 +68,62 @@ let sync_pci_hidden ~__context ~pgpu ~pci =
+        with Scanf.Scan_failure _ | Failure _ | End_of_file -> false
+    ) then (
+      let physfn_addr = String.sub linktarget 3 12 in
+      let expr = (let open Db_filter_types in Eq (Field "pci_id", Literal physfn_addr)) in
```
Slightly nicer is this:

    let expr = Db_filter_types.(Eq (Field "pci_id", Literal physfn_addr)) in

robhoes commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -68,15 +68,62 @@ let sync_pci_hidden ~__context ~pgpu ~pci =
+    ) then (
+      let physfn_addr = String.sub linktarget 3 12 in
+      let expr = (let open Db_filter_types in Eq (Field "pci_id", Literal physfn_addr)) in
+      ( match Db.PCI.get_records_where ~__context ~expr with (* Expect exactly one *)
```
No, because the DB contains PCIs from all hosts in the pool! The expression must filter by localhost.

thomassa replied:
Good catch.

thomassa commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -307,3 +354,25 @@ let disable_dom0_access ~__context ~self =
+  Xapi_pci.update_pcis ~__context ~host:localhost;
+  (* Potential optimisation: make update_pcis return a value telling whether
+   * it changed anything, and stop here if it did not. *)
+  List.iter
```
It would be harmless, but would do a fair bit of needless work, and doing it this way is still quite simple.
There will never be any need to create or update any pgpu DB objects here: all we want to do is create a pci DB objects for each VF pci device, and set its `physical_function` field (which will automatically cause the `virtual_functions` field of its PF pci DB object to be updated).

robhoes replied:
Sure, let's leave it, since it'll be on the VM.start code path.

robhoes commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -307,3 +354,25 @@ let disable_dom0_access ~__context ~self =
+   * it changed anything, and stop here if it did not. *)
+  List.iter
+    (fun pci_ref -> ignore (is_pgpu ~__context ~localhost pci_ref))
+    (Db.PCI.get_all ~__context)
```
We could optimise this by getting only the PCIs for the localhost, using `get_records_where`. And pass the record to `is_pgpu` and beyond, so save some further DB calls.

thomassa commented:
I believe I've addressed all the review-comments; if you're happy with this, I'll squash and merge.

But also, I have some work-in-progress to remove the `host` or `localhost` parameters that have to be the local host: those functions should use `Helpers.get_localhost ~__context` instead of accepting a parameter.

robhoes commented on ocaml/xapi/xapi_pgpu.ml:
```diff
@@ -68,15 +68,61 @@ let sync_pci_hidden ~__context ~pgpu ~pci =
-        Db.PCI.get_host ~__context ~self = host
-        && Xapi_pci.(is_class_of_kind Display_controller (int_of_id class_id))
-      ) (Db.PCI.get_all ~__context) in
+    (* Important side-effects in is_local_pgpu *)
```
You mean `is_pgpu`?

robhoes commented:
Looks good. I think you just made a potentially confusing typo in two comments.

thomassa commented:
Updated.
